tests/core/test_signal.py

```diff
@@ -81,6 +83,99 @@ async def test_signal_connect_fails_if_different_backend_but_same_by_value():
+    caplog.set_level(logging.DEBUG)
+    int_mock_backend = MockSignalBackend(int)
+    original_connect = int_mock_backend.connect
+
+    async def new_connect(timeout=1):
+        await asyncio.sleep(0.1)
+        await original_connect(timeout=timeout)
+
+    int_mock_backend.connect = new_connect
+    signal = Signal(int_mock_backend)
+    await signal.connect()  # no need to pass the mock parameter as already specified
+    assert signal._backend.datatype == int
+    await signal.connect()  # no need to pass the mock parameter as already specified
+    response = f"Reusing previous connection to {signal.source}"
+    assert response in caplog.text
+
+
+async def test_signal_connects_with_force_reconnect(caplog):
+    caplog.set_level(logging.DEBUG)
+    signal = Signal(MockSignalBackend(int))
+    await signal.connect()  # no need to pass the mock parameter as already specified
+    assert signal._backend.datatype == int
+    await signal.connect(
+        force_reconnect=True
+    )  # no need to pass the mock parameter as already specified
+    response = f"Connecting to {signal.source}"
+    assert response in caplog.text
+
+
+@pytest.mark.parametrize(
+    "first, second",
+    [(True, False), (True, False)],
+)
+async def test_rejects_reconnect_when_connects_have_diff_mock_status(
+    caplog, first, second
+):
+    caplog.set_level(logging.DEBUG)
+    signal = Signal(MockSignalBackend(int))
+    await signal.connect(mock=first)
+    assert signal._backend.datatype == int
+    await signal.connect(mock=second)
+    response = f"Connecting to {signal.source}"
+    assert caplog.text.__contains__(response)
+
+
+async def test_signal_lazily_connects(RE):
+    mock_signal_rw = soft_signal_rw(
+        int, "pva://this is an initial value not a pva pv", name="mock_signal"
+    )
+    await mock_signal_rw.connect(mock=False)
+    RE(ensure_connected(mock_signal_rw, mock=True))
+    assert (
+        mock_signal_rw._connect_task
+        and mock_signal_rw._connect_task.done()
+        and not mock_signal_rw._connect_task.exception()
+    )
+
+
+async def test_signal_mock_and_back_again(RE):
+    mock_signal_rw = soft_signal_rw(int, 0, name="mock_signal")
+    assert not mock_signal_rw._connect_task
+    await mock_signal_rw.connect(mock=False)
+    assert isinstance(mock_signal_rw._backend, SoftSignalBackend)
+    assert mock_signal_rw._connect_task
+    await mock_signal_rw.connect(mock=True)
+    assert isinstance(mock_signal_rw._backend, MockSignalBackend)
+
+
+async def test_connect_fails_then_next_connect_succeeds():
+    mock_signal_rw = soft_signal_rw(int, 0, name="mock_signal")
+    assert not mock_signal_rw._connect_task
+    await mock_signal_rw.connect(mock=False)
+    assert isinstance(mock_signal_rw._backend, SoftSignalBackend)
+    assert mock_signal_rw._connect_task
+    # todo is explicit disconnection needed?
+    await mock_signal_rw.connect(mock=True)
+    assert isinstance(mock_signal_rw._backend, MockSignalBackend)
```

Comment:
These should both fail, they're not because you aren't setting `_previous_connect_was_mock`.

Reply:
yes those are the two remaining ones that fails. 
`test_connect_fails_then_next_connect_succeeds` and `test_signal_mock_and_back_again`

we either need to delete the tests or change their definition in the test name and then the code

Reply:
Well `test_signal_mock_and_back_again` should check that the above behavior raises an error, which I think you've done in a different test - if so we can delete.

As said previously, `test_connect_fails_then_next_connect_succeeds` doesn't test what the title would imply. `soft_signal_rw` will connect fine if `mock=True` or `mock=False`.

You could create a new connect function just for this test which will fail on the first `connect`, but pass after.

``` python
signal = pva_signal_backend(int, "pva://some_pv_that_doesnt_exist", initial_value="")
original_connect_method = signal.connect
connect_attempts = 0
async def connect(*args, **kwargs):
    # Will cause an error on the first connect call, then pass after
    if connect_attempts == 0:
        connect_attempts += 1
        await original_connect(*args, **kwargs)
signal.connect = connect
```

Reply:
> You could create a new connect function just for this test 


I think that's uphill and excessive

Reply:
outdated, the tests got deleted

Reply:
But this is the entire reason that mock was switching in those tests that you renamed - it wasn't to test the behavior of the connect function when using mock utilities... It was to see if you can try and fail to connect, then try again and have it work.